File: algorithms/sorting/quick-sort/quick_sort.cpp
#include <utility>
#include <vector>

std::size_t partition(std::vector<int>& array, std::size_t start, std::size_t end) {
    const int pivot = array[end];
    std::size_t boundary = start;

    for (std::size_t scan = start; scan < end; ++scan) {
        if (array[scan] <= pivot) {
            std::swap(array[boundary], array[scan]);
            ++boundary;
        }
    }

    std::swap(array[boundary], array[end]);
    return boundary;
}

void sortRange(std::vector<int>& array, std::size_t start, std::size_t end) {
    if (start >= end) return;
    const std::size_t pivotIndex = partition(array, start, end);
    if (pivotIndex > 0) sortRange(array, start, pivotIndex - 1);
    sortRange(array, pivotIndex + 1, end);
}

std::vector<int> quickSort(const std::vector<int>& numbers) {
    std::vector<int> array = numbers;
    if (!array.empty()) sortRange(array, 0, array.size() - 1);
    return array;
}

File: algorithms/sorting/quick-sort/QuickSort.java
import java.util.Arrays;

public final class QuickSort {
    private QuickSort() {
    }

    public static int[] quickSort(int[] numbers) {
        int[] array = Arrays.copyOf(numbers, numbers.length);
        sortRange(array, 0, array.length - 1);
        return array;
    }

    private static void sortRange(int[] array, int start, int end) {
        if (start >= end) return;
        int pivotIndex = partition(array, start, end);
        sortRange(array, start, pivotIndex - 1);
        sortRange(array, pivotIndex + 1, end);
    }

    private static int partition(int[] array, int start, int end) {
        int pivot = array[end];
        int boundary = start;
        for (int scan = start; scan < end; scan++) {
            if (array[scan] <= pivot) {
                int temporary = array[boundary];
                array[boundary] = array[scan];
                array[scan] = temporary;
                boundary++;
            }
        }
        int temporary = array[boundary];
        array[boundary] = array[end];
        array[end] = temporary;
        return boundary;
    }
}
